# Capstone project
## 1. Scraping Toronto neighborhood data from wikipedia

install pacakges

In [11]:
#!pip install beautifulsoup4
#!pip install lxml
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

from IPython.display import display_html
import pandas as pd
import numpy as np
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

import folium # plotting library
from bs4 import BeautifulSoup
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


Toronto's neighbor list is available in wikipedia page 'List of postal codes of Canada'. Get the list using web-scraping pacakge 'beautiful soup'. 
Once we get content of the page, html structure are stored. We can find the table using `find` function and extract table content.

In [12]:
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup=BeautifulSoup(source,'lxml')
print(soup.title)

<title>List of postal codes of Canada: M - Wikipedia</title>


- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area.These two rows will be combined into one row with the neighborhoods separated with a comma.
- If a cell has a borough but a Not assigned  neighborhood, then the neighborhood will be the same as the borough.

In [13]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    #Ignore cells with a borough that is Not assigned.
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [14]:
df.shape

(103, 3)

103 neighbors are listed here.

## 2. Get coodinates of each neighborhoodsg
To use foursqure API, postal code is not enough and we need coordinates of latitude and longitude. To get thoes, first I tried out geocoder package like below, however, found it is time consuming and as coordinates are provided as csv in instruction, I decided to take it.

``` python 

#!pip install geocoder
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format('M5G'))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

```

In [15]:
coords = pd.read_csv('Geospatial_Coordinates.csv')
coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [16]:
coords.columns = ['PostalCode', 'Latitude', 'Longitude']
df = pd.merge(df, coords, 'outer', on='PostalCode')
df

PostalCode                Borough  ...   Latitude  Longitude
0          M3A             North York  ...  43.753259 -79.329656
1          M4A             North York  ...  43.725882 -79.315572
2          M5A       Downtown Toronto  ...  43.654260 -79.360636
3          M6A             North York  ...  43.718518 -79.464763
4          M7A           Queen's Park  ...  43.662301 -79.389494
..         ...                    ...  ...        ...        ...
98         M8X              Etobicoke  ...  43.653654 -79.506944
99         M4Y       Downtown Toronto  ...  43.665860 -79.383160
100        M7Y  East Toronto Business  ...  43.662744 -79.321558
101        M8Y              Etobicoke  ...  43.636258 -79.498509
102        M8Z              Etobicoke  ...  43.628841 -79.520999

[103 rows x 5 columns]

## 3. Explor and clustering neighborhoods
Create map around toronto to get initial impression of geography

In [17]:
# Coordinate of downtown toronto
latitude = df[df['Borough'] =='Downtown Toronto'].iloc[0].Latitude
longitude = df[df['Borough'] =='Downtown Toronto'].iloc[0].Longitude

# create map of New York using latitude and longitude values
map_canada = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_canada)  
    
map_canada

Based on coordinates table, get nearby venue using foursqure API. The function used for New York in previous section can be applied.  

In [18]:
#Credential setting for foursqure developer API
CLIENT_ID = '4CXJVQTJNPBIFY5413VOP2WRVQK0EGJ400OOJI2QJU1HCBB4' # your Foursquare ID
CLIENT_SECRET = 'I3ECCDVCHAW2MDLJNISPQJYOU0ZYJGEK0CWECVOEYX1HWX5K' # your Foursquare Secret
ACCESS_TOKEN = 'CYCAIMSPDWIJBXNN5HVCNN5PBGZIMUHQFBTPG354MGFQ5EQ0' # your FourSquare Access Token
VERSION = '20180604'
LIMIT = 30

In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [21]:
toronto_venues.head()

Neighborhood  ...         Venue Category
0         Parkwoods  ...                   Park
1         Parkwoods  ...   Fast Food Restaurant
2         Parkwoods  ...      Food & Drink Shop
3  Victoria Village  ...           Hockey Arena
4  Victoria Village  ...  Portuguese Restaurant

[5 rows x 7 columns]

Now all venues were obtained. Check how many venues are in each neighbors

In [22]:
toronto_venues.groupby('Neighborhood').Venue.count()

Neighborhood
Agincourt                                           4
Alderwood, Long Branch                              7
Bathurst Manor, Wilson Heights, Downsview North    19
Bayview Village                                     4
Bedford Park, Lawrence Manor East                  25
                                                   ..
Willowdale, Newtonbrook                             1
Woburn                                              5
Woodbine Heights                                    6
York Mills West                                     2
York Mills, Silver Hills                            1
Name: Venue, Length: 101, dtype: int64

In [23]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 240 uniques categories.


To cluster venues, use one-hot encording

In [24]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  ...  Wings Joint  Women's Store
0            0                  0  ...            0              0
1            0                  0  ...            0              0
2            0                  0  ...            0              0
3            0                  0  ...            0              0
4            0                  0  ...            0              0

[5 rows x 240 columns]

In [25]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  ...  Women's Store
0                                          Agincourt  ...            0.0
1                             Alderwood, Long Branch  ...            0.0
2    Bathurst Manor, Wilson Heights, Downsview North  ...            0.0
3                                    Bayview Village  ...            0.0
4                  Bedford Park, Lawrence Manor East  ...            0.0
..                                               ...  ...            ...
96                           Willowdale, Newtonbrook  ...            0.0
97                                            Woburn  ...            0.0
98                                  Woodbine Heights  ...            0.0
99                                   York Mills West  ...            0.0
100                         York Mills, Silver Hills  ...            0.0

[101 rows x 240 columns]

In [26]:
novenue = [i for i in df.Neighborhood if i not in toronto_grouped.Neighborhood.values.tolist()]
novenue

['Islington Avenue', 'Upper Rouge']

Islington Avenue and upper rouge have no venue data. Thoes will be excluded from clustering.

### Common venues of each neighbors

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  ... 10th Most Common Venue
0                                        Agincourt  ...     Dim Sum Restaurant
1                           Alderwood, Long Branch  ...           Dessert Shop
2  Bathurst Manor, Wilson Heights, Downsview North  ...            Supermarket
3                                  Bayview Village  ...         Discount Store
4                Bedford Park, Lawrence Manor East  ...      Indian Restaurant

[5 rows x 11 columns]

### K-Mean Clustering
Using grouped dataframe, cluster neighbor into 6 group using k-means, like as New York neighboer clustering.

In [32]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 5], dtype=int32)

In [35]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood',how='inner')

toronto_merged.head() # check the last columns!

PostalCode           Borough  ... 9th Most Common Venue  10th Most Common Venue
0        M3A        North York  ...                 Diner      Dim Sum Restaurant
1        M4A        North York  ...         Deli / Bodega           Women's Store
2        M5A  Downtown Toronto  ...                   Pub      Mexican Restaurant
3        M6A        North York  ...     Accessories Store            Airport Gate
4        M7A      Queen's Park  ...         Smoothie Shop       College Cafeteria

[5 rows x 16 columns]

### Visualize clustering result

In [30]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In toronto case, most of neighbors (63 borough) are fell into same cluster, meaning others have some unique venues. As we learned previouse course, K-mean is not suitable algorithm for this case because outlier point is also assighned to some cluster. However, it depends on the perpouse that which algorithm is the best, if we need to idnentify similar boroughs in wide rage, k-mean cluster would also good choice.